generate/src/contract/events.rs

```diff
@@ -132,19 +304,76 @@ mod tests {
 
         assert_quote!(expand_filter(&event).unwrap(), {
             /// Generated by `ethcontract`.
-            pub fn transfer(&self) -> self::ethcontract::dyns::DynEventBuilder<(
-                self::ethcontract::Address,
-                self::ethcontract::Address,
-                self::ethcontract::U256
-            )> {
+            pub fn transfer(&self) -> self::ethcontract::dyns::DynEventBuilder<self::events::Transfer> {
                 self.instance.event(#signature).expect("generated event filter")
             }
         });
     }
 
     #[test]
+    fn expand_data_struct_value() {
+        let event = Event {
+            name: "Foo".into(),
+            inputs: vec![
+                EventParam {
+                    name: "a".into(),
+                    kind: ParamType::Bool,
+                    indexed: false,
+                },
+                EventParam {
+                    name: String::new(),
+                    kind: ParamType::Address,
+                    indexed: false,
+                },
+            ],
+            anonymous: false,
+        };
+
+        let name = expand_struct_name(&event);
+        let params = expand_params(&event).unwrap();
+        let (definition, construction) = expand_data_struct(&name, &params);
+
+        assert_quote!(definition, {
+            struct Foo {
+                pub a: bool,
+                pub p1: self::ethcontract::Address,
+            }
+        });
+        assert_quote!(construction, { Foo { a, p1 } });
+    }
+
+    #[test]
+    fn expand_data_tuple_value() {
+        let event = Event {
+            name: "Foo".into(),
+            inputs: vec![
+                EventParam {
+                    name: String::new(),
+                    kind: ParamType::Bool,
+                    indexed: false,
+                },
+                EventParam {
+                    name: String::new(),
+                    kind: ParamType::Address,
+                    indexed: false,
+                },
+            ],
+            anonymous: false,
```

Comment:
What is this "anonymous" event component I have never heard of until now?

Reply:
Solidity supports anonymous events. I also had no idea until I started doing some research for this feature.

Reply:
Anonymous events don't include the event signature in topic 0, they basically just put the first indexed field into topic 0, so you lose the ability to filter on specific kinds of events from an event signature. The advantage is you get one extra topic to filter on.

I imagine this can be useful for contracts that only have a single event.